### ISSUE TITLE
Is there any evidence that softmax one can takes advantages over normal softmax?

### ISSUE THREAD

**Issue by ZGCTroy** [kyegomez/AttentionIsOFFByOne#6]

**Comment by snoop2head** [comment]
**Personal experimentation reported decreased performance and unstable training process with the softmax_one function with DeiT3 trained on ImageNet1K .**

### Training Log
- Followed training ViT-Small configuration of [DeiT III: Revenge of the ViT](https://arxiv.org/pdf/2204.07118.pdf).
- ViT-S with normal softmax: 80.8% validation set top1 accuracy.
- ViT-S with softmax_one: 80.7% validation set top1 accuracy + fluctuations in loss and accuracy.

![Screenshot 2023-10-31 at 1 30 30 PM](https://github.com/kyegomez/AttentionIsOFFByOne/assets/28736695/9d03e0df-f845-466b-bd40-eff2cff2ccaf)

### Implementation using Jax

```python3
def softmax_one(
    x, 
    axis: Optional[Union[int, Tuple[int, ...]]] = -1,
    where: Optional[Array] = None,
    initial: Optional[Array] = None) -> Array:
    
    # subtract the max for stability and compute exponentials
    x_max = jnp.max(x, axis, where=where, initial=initial, keepdims=True)
    exp_x = jnp.exp(x - x_max)
    #compute softmax values and add on in the denominator
    result = exp_x / (1 + jnp.sum(exp_x, axis=axis, keepdims=True))
    if where is not None:
        result = jnp.where(where, result, 0)
    return result
```

Please let me know if I've done any mistakes. 
I will run same experimentations again if that's the case.

### Personal Thoughts
I think adding zero token for the attention score aggravates attention collapse problem in the Transformer architecture. Attention collapse problem is stated in [DeepViT: Towards Deeper Vision Transformer](https://arxiv.org/pdf/2103.11886.pdf) as following:

> We empirically observed that as the depth of ViTs increases, the attention maps, used for aggregating the features for each transformer block, tend to be overly similar after certain layers, which makes the representations stop evolving after certain layers. We name this specific issue as attention collapse. This indicates that as the ViT goes deeper, the self-attention mechanism becomes less effective for generating diverse attentions to capture rich representations.

[Query-Key Normalization for Transformers(EMNLP 2020)](https://aclanthology.org/2020.findings-emnlp.379.pdf) suggested that their method of L2 normalization on Queries and Keys lead to diffuse attention patterns and increased performance. Their visualization of attention map with/without the normalization is given below.

![Screenshot 2023-10-31 at 1 48 45 PM](https://github.com/kyegomez/AttentionIsOFFByOne/assets/28736695/a2f2773e-01d8-4f7c-b168-3b3f634ec0bc)

Another paper [Scaling Vision Transformers to 22 Billion Parameters(ICML 2023)](https://arxiv.org/abs/2302.05442)  also proposed that instability of training large-scale transformers can be controlled by normalization on Queries and Keys. The graph below show that attention score almost becomes one-hot vector(5th graph), preceding the divergence of training loss(1st graph). 

![Screenshot 2023-10-31 at 1 49 56 PM](https://github.com/kyegomez/AttentionIsOFFByOne/assets/28736695/5344d755-caf8-4f5d-895b-43768283df89)

On my personal opinion, dense/diverse attention scores benefits the training rather than attention maps consisted with zeros and ones. Thus I think softmax_one may comparably aggravate attention collapse since previous legacy softmax couldn't reach zero value for the attention score, but softmax_one can.
